# Pandas part 2!

<img src="panda_2.gif"> 

[gif source](https://media1.giphy.com/media/EatwJZRUIv41G/giphy.gif?cid=6c09b952i55b90ss4xr003k2a9m5u7ix541wglk68og6g3re&ep=v1_gifs_search&rid=giphy.gif&ct=g)


`pandas` is a Python package that is widely used for managing/cleaning tabular data.,

In this example, we will continue looking at data from police stops - where each row is 1 stop, and each column contains so desciption of that stop - like the time the stop look place, the outcome of the stop, etc.

We will continue from the work we started last week - so we will jump to the "Check null values across columns section..."! This week we'll be using `.dropna()`, `.sum()`, and `.value_counts()` functions, and practicing logical indexing!



And as always... *remember the using documentation is a critical data science skill* and part of the process!

The dataset here comes from [Kaggle](https://www.kaggle.com/datasets/melihkanbay/police) and is intended to be used to practice using pandas.

In [1]:
# if using Google Collab - uncomment the following lines and run them. 
#
# You'll also need to add the entire path of your data file in your google drive.
#     > You can use the 'files' tab on the left to find your data file, then right click and copy path to get the entire path
#     > Pass the path in "" inside the parameters when you read in the data.


# from google.colab import drive
# drive.mount('/content/drive/',  force_remount=True)

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ali-rivera/Python-Support-Hours/blob/main/Week3/Week3_BlankResources.ipynb)

In [ ]:
import pandas as pd

## Read in data

In [ ]:
data = pd.read_csv("police.csv")
data

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0      2005-01-02     01:55          NaN             M          1985.0   
1      2005-01-18     08:15          NaN             M          1965.0   
2      2005-01-23     23:15          NaN             M          1972.0   
3      2005-02-20     17:15          NaN             M          1986.0   
4      2005-03-14     10:00          NaN             F          1984.0   
...           ...       ...          ...           ...             ...   
91736  2015-12-31     20:27          NaN             M          1986.0   
91737  2015-12-31     20:35          NaN             F          1982.0   
91738  2015-12-31     20:45          NaN             M          1992.0   
91739  2015-12-31     21:42          NaN             M          1993.0   
91740  2015-12-31     22:46          NaN             M          1959.0   

       driver_age driver_race                   violation_raw  \
0            20.0       White                        Speeding   
1            40.0       White                        Speeding   
2            33.0       White                        Speeding   
3            19.0       White                Call for Service   
4            21.0       White                        Speeding   
...           ...         ...                             ...   
91736        29.0       White                        Speeding   
91737        33.0       White  Equipment/Inspection Violation   
91738        23.0       White         Other Traffic Violation   
91739        22.0       White                        Speeding   
91740        56.0    Hispanic                        Speeding   

              violation  search_conducted search_type   stop_outcome  \
0              Speeding             False         NaN       Citation   
1              Speeding             False         NaN       Citation   
2              Speeding             False         NaN       Citation   
3                 Other             False         NaN  Arrest Driver   
4              Speeding             False         NaN       Citation   
...                 ...               ...         ...            ...   
91736          Speeding             False         NaN        Warning   
91737         Equipment             False         NaN        Warning   
91738  Moving violation             False         NaN        Warning   
91739          Speeding             False         NaN       Citation   
91740          Speeding             False         NaN       Citation   

      is_arrested stop_duration  drugs_related_stop  
0           False      0-15 Min               False  
1           False      0-15 Min               False  
2           False      0-15 Min               False  
3            True     16-30 Min               False  
4           False      0-15 Min               False  
...           ...           ...                 ...  
91736       False      0-15 Min               False  
91737       False      0-15 Min               False  
91738       False      0-15 Min               False  
91739       False      0-15 Min               False  
91740       False      0-15 Min               False  

[91741 rows x 15 columns]

## Drop Redundant Columns

First lets take a look and decide if there are any columns we just don't want. It looks like `driver_age_raw` may be the driver's birth year and age has been added (or calculated) as it's own column. Since this information is redundant, I think we can go ahead and drop `driver_age_raw`.

We can do this using the [.drop()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) function.

In [ ]:
clean_data = data.drop(columns = "driver_age_raw")
clean_data


stop_date stop_time  county_name driver_gender  driver_age  \
0      2005-01-02     01:55          NaN             M        20.0   
1      2005-01-18     08:15          NaN             M        40.0   
2      2005-01-23     23:15          NaN             M        33.0   
3      2005-02-20     17:15          NaN             M        19.0   
4      2005-03-14     10:00          NaN             F        21.0   
...           ...       ...          ...           ...         ...   
91736  2015-12-31     20:27          NaN             M        29.0   
91737  2015-12-31     20:35          NaN             F        33.0   
91738  2015-12-31     20:45          NaN             M        23.0   
91739  2015-12-31     21:42          NaN             M        22.0   
91740  2015-12-31     22:46          NaN             M        56.0   

      driver_race                   violation_raw         violation  \
0           White                        Speeding          Speeding   
1           White                        Speeding          Speeding   
2           White                        Speeding          Speeding   
3           White                Call for Service             Other   
4           White                        Speeding          Speeding   
...           ...                             ...               ...   
91736       White                        Speeding          Speeding   
91737       White  Equipment/Inspection Violation         Equipment   
91738       White         Other Traffic Violation  Moving violation   
91739       White                        Speeding          Speeding   
91740    Hispanic                        Speeding          Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
91736             False         NaN        Warning       False      0-15 Min   
91737             False         NaN        Warning       False      0-15 Min   
91738             False         NaN        Warning       False      0-15 Min   
91739             False         NaN       Citation       False      0-15 Min   
91740             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
91736               False  
91737               False  
91738               False  
91739               False  
91740               False  

[91741 rows x 14 columns]

## Rename for easier indexing

Let's rename some of the the columns to be a bit shorter while still being descriptive. We can use the [.rename()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html) function to do this.

In [ ]:
clean_data.columns

Index(['stop_date', 'stop_time', 'county_name', 'driver_gender', 'driver_age',
       'driver_race', 'violation_raw', 'violation', 'search_conducted',
       'search_type', 'stop_outcome', 'is_arrested', 'stop_duration',
       'drugs_related_stop'],
      dtype='object')

In [ ]:
cols_dict = {'stop_date': 'date',
             'stop_time': 'time',
             'county_name': 'county',
             'driver_gender': 'gender',
             'driver_age': 'age',
             'driver_race': 'race',
             'stop_outcome': 'outcome',
             'is_arrested': 'arrested',
             'drugs_related_stop': 'drug_related'}
             
clean_data.rename(columns = cols_dict, inplace=True)
clean_data

date   time  county gender   age      race  \
0      2005-01-02  01:55     NaN      M  20.0     White   
1      2005-01-18  08:15     NaN      M  40.0     White   
2      2005-01-23  23:15     NaN      M  33.0     White   
3      2005-02-20  17:15     NaN      M  19.0     White   
4      2005-03-14  10:00     NaN      F  21.0     White   
...           ...    ...     ...    ...   ...       ...   
91736  2015-12-31  20:27     NaN      M  29.0     White   
91737  2015-12-31  20:35     NaN      F  33.0     White   
91738  2015-12-31  20:45     NaN      M  23.0     White   
91739  2015-12-31  21:42     NaN      M  22.0     White   
91740  2015-12-31  22:46     NaN      M  56.0  Hispanic   

                        violation_raw         violation  search_conducted  \
0                            Speeding          Speeding             False   
1                            Speeding          Speeding             False   
2                            Speeding          Speeding             False   
3                    Call for Service             Other             False   
4                            Speeding          Speeding             False   
...                               ...               ...               ...   
91736                        Speeding          Speeding             False   
91737  Equipment/Inspection Violation         Equipment             False   
91738         Other Traffic Violation  Moving violation             False   
91739                        Speeding          Speeding             False   
91740                        Speeding          Speeding             False   

      search_type        outcome arrested stop_duration  drug_related  
0             NaN       Citation    False      0-15 Min         False  
1             NaN       Citation    False      0-15 Min         False  
2             NaN       Citation    False      0-15 Min         False  
3             NaN  Arrest Driver     True     16-30 Min         False  
4             NaN       Citation    False      0-15 Min         False  
...           ...            ...      ...           ...           ...  
91736         NaN        Warning    False      0-15 Min         False  
91737         NaN        Warning    False      0-15 Min         False  
91738         NaN        Warning    False      0-15 Min         False  
91739         NaN       Citation    False      0-15 Min         False  
91740         NaN       Citation    False      0-15 Min         False  

[91741 rows x 14 columns]

## Correct dtypes

Before we go any further, let's look at the data type of each column and make sure they are appropriate for the information stored there. There are a few functions we can use to do this, [.info()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) gives us the most information. [.dtypes](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html) also gives us a list of the data types, if that's all we want.

In [ ]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   date              91741 non-null  object 
 1   time              91741 non-null  object 
 2   county            0 non-null      float64
 3   gender            86406 non-null  object 
 4   age               86120 non-null  float64
 5   race              86408 non-null  object 
 6   violation_raw     86408 non-null  object 
 7   violation         86408 non-null  object 
 8   search_conducted  91741 non-null  bool   
 9   search_type       3196 non-null   object 
 10  outcome           86408 non-null  object 
 11  arrested          86408 non-null  object 
 12  stop_duration     86408 non-null  object 
 13  drug_related      91741 non-null  bool   
dtypes: bool(2), float64(2), object(10)
memory usage: 8.6+ MB


Notice the "RangeIndex" tells use how many rows we have (91741) and "Data Columns" tells us we have 14 columns(after dropping `driver_age_raw`). The "Non-Null Count" column in this outpust shows us that several columns have null values for several rows. We'll keep that in mind for later. For now, let's look a the type of each column.

`stop_date` and `stop_time` are stored as objects, but storing these as a datetime datatype may serve us better. Let's make a new column called `date_time` and store the date AND time in there using the [.to_datetime()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html#pandas.to_datetime) function.

In [ ]:
clean_data['date_time'] = pd.to_datetime(clean_data.date + " " + clean_data.time)

In [ ]:
clean_data.date_time

0       2005-01-02 01:55:00
1       2005-01-18 08:15:00
2       2005-01-23 23:15:00
3       2005-02-20 17:15:00
4       2005-03-14 10:00:00
                ...        
91736   2015-12-31 20:27:00
91737   2015-12-31 20:35:00
91738   2015-12-31 20:45:00
91739   2015-12-31 21:42:00
91740   2015-12-31 22:46:00
Name: date_time, Length: 91741, dtype: datetime64[ns]

In [ ]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              91741 non-null  object        
 1   time              91741 non-null  object        
 2   county            0 non-null      float64       
 3   gender            86406 non-null  object        
 4   age               86120 non-null  float64       
 5   race              86408 non-null  object        
 6   violation_raw     86408 non-null  object        
 7   violation         86408 non-null  object        
 8   search_conducted  91741 non-null  bool          
 9   search_type       3196 non-null   object        
 10  outcome           86408 non-null  object        
 11  arrested          86408 non-null  object        
 12  stop_duration     86408 non-null  object        
 13  drug_related      91741 non-null  bool          
 14  date_time         9174

Now let's take a look at `gender`, `race`, `violation`, `outcome`, `search_type`, and `stop_duration`, which I suspect may be able to be stored as categories...

We can use the [.value_counts()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html) function for this, which lists out all the values in a column and the count of how many cells hold each value.

In [ ]:
cat_cols = ["gender", "race", "violation", "outcome", "search_type", "stop_duration"]

for col in cat_cols:
    print(clean_data[col].value_counts(), "\n")

gender
M    62895
F    23511
Name: count, dtype: int64 

race
White       62158
Black       12244
Hispanic     9507
Asian        2259
Other         240
Name: count, dtype: int64 

violation
Speeding               48463
Moving violation       16224
Equipment              11020
Other                   4317
Registration/plates     3432
Seat belt               2952
Name: count, dtype: int64 

outcome
Citation            77006
Warning              5294
Arrest Driver        2571
N/D                   590
No Action             589
Arrest Passenger      358
Name: count, dtype: int64 

search_type
Incident to Arrest                                          1219
Probable Cause                                               891
Inventory                                                    220
Reasonable Suspicion                                         197
Protective Frisk                                             161
Incident to Arrest,Inventory                                 129
Incident to Ar

Looks like all of these except `search_type` would be better as categories, lets retype them as such using the [.astype()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.astype.html) function.

In [ ]:
cat_cols.remove("search_type") #remove search_type from the list
clean_data[cat_cols] = clean_data[cat_cols].astype('category') #cast all columns in cat_cols list as categories and save the results in the og column(s)

In [ ]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              91741 non-null  object        
 1   time              91741 non-null  object        
 2   county            0 non-null      float64       
 3   gender            86406 non-null  category      
 4   age               86120 non-null  float64       
 5   race              86408 non-null  category      
 6   violation_raw     86408 non-null  object        
 7   violation         86408 non-null  category      
 8   search_conducted  91741 non-null  bool          
 9   search_type       3196 non-null   object        
 10  outcome           86408 non-null  category      
 11  arrested          86408 non-null  object        
 12  stop_duration     86408 non-null  category      
 13  drug_related      91741 non-null  bool          
 14  date_time         9174

It looks like `arrested` is stored as an object but seems to be True/False. Let's take a look at the values and see if we can store it as a bool...

In [ ]:
clean_data.arrested.value_counts()

arrested
False    83479
True      2929
Name: count, dtype: int64

In [ ]:
clean_data.arrested = clean_data.arrested.astype("bool") #cast & save arrested as a boolean since it is True/False

In [ ]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              91741 non-null  object        
 1   time              91741 non-null  object        
 2   county            0 non-null      float64       
 3   gender            86406 non-null  category      
 4   age               86120 non-null  float64       
 5   race              86408 non-null  category      
 6   violation_raw     86408 non-null  object        
 7   violation         86408 non-null  category      
 8   search_conducted  91741 non-null  bool          
 9   search_type       3196 non-null   object        
 10  outcome           86408 non-null  category      
 11  arrested          91741 non-null  bool          
 12  stop_duration     86408 non-null  category      
 13  drug_related      91741 non-null  bool          
 14  date_time         9174

## Check null values across columns - decide what to keep and what to get rid of

All our column tpyes look good now! Let's take a look at `county`, since it looks like that has 0 non-null values...

It looks like we are missing some information in `gender`, `age`, `race`, `violation_raw`, `violation`, `outcome`, and `stop_duration`. Let's use the [.dropna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) function to drop rows that are missing values in those columns.

Now let's take a look at `search_type`, which I hypothesize is null when there is no search. Let's see how many `search_conducted` fields are True and if that matches our number of `search_type` non-null fields. There's a few ways do do this...

## Seperate search & non-search stops

So it looks like when a search is conducted we have search type. Maybe we want to seperate these into seperate dataframes, one for search stops and one for non-search stops...

*Note: .reset_index() resets the index to start at 0 and increment by 1. The drop = True arguement keeps us from having 2 indices. Try removing each of these to see what happens.*

## Next Steps

Try some of the following (open ended) things on your own:
- Pick a demographic (a certain age range, a race, a gender) and a metric (stop_duration, time of day, outcome, etc). Try to filter on each of these things in 2 different ways.
- `age` is a numeric value. We may want to see the range of ages we have in our dataset. Find a pandas function to get summary statistics on the age variable.
    - Bonus: make a histogram of the age distribution!
- Make age categorical instead of numerical. Pick any number of categories you'd like and the cutoffs. Hint: this is called 'binning' and there is a function to do it! You could also do this with `time`...
    - suggestions: 3 categories - young, middle_aged, and old; by 10s: 10-19, 20-19, 30-39, etc.
- Pick a demographic and *annonomize* it. For example: replace the `gender` category with A/B instead of M/F.
